packages/client/lib/miner/miner.ts

```diff
@@ -0,0 +1,258 @@
+  }
+
+  /**
+   * Sets the timeout for the next block assembly
+   */
+  private async queueNextAssembly(timeout?: number) {
+    if (this._nextAssemblyTimeoutId) {
+      clearTimeout(this._nextAssemblyTimeoutId)
+    }
+    timeout = timeout ?? this.period
+    if (this.config.chainCommon.consensusType() === ConsensusType.ProofOfAuthority) {
+      // EIP-225 spec: If the signer is out-of-turn,
+      // delay signing by rand(SIGNER_COUNT * 500ms)
+      const [signerAddress] = this.config.accounts[0]
+      const { blockchain } = this.synchronizer.execution.vm
+      const inTurn = await blockchain.cliqueSignerInTurn(signerAddress)
+      if (!inTurn) {
+        const signerCount = blockchain.cliqueActiveSigners().length
+        timeout += Math.random() * signerCount * 500
+      }
+    }
+    this._nextAssemblyTimeoutId = setTimeout(this.assembleBlock.bind(this), timeout)
+  }
+
+  /**
+   * Sets the next block assembly to latestBlock.timestamp + period
+   */
+  private async chainUpdated() {
+    const latestBlockHeader = this.latestBlockHeader()
+    const target = latestBlockHeader.timestamp.muln(1000).addn(this.period).sub(new BN(Date.now()))
+    const timeout = BN.max(new BN(0), target).toNumber()
+    this.config.logger.debug(
+      `Miner: Chain updated with block ${latestBlockHeader.number.toNumber()}. Queuing next block assembly in ${Math.round(
+        timeout / 1000
+      )}s`
+    )
+    await this.queueNextAssembly(timeout)
+  }
+
+  /**
+   * Start miner
+   */
+  start(): boolean {
+    if (!this.config.mine || this.running) {
+      return false
+    }
+    this.running = true
+    this.config.events.on(Event.CHAIN_UPDATED, this.chainUpdated.bind(this))
+    void this.queueNextAssembly() // void operator satisfies eslint rule for no-floating-promises
+    this.config.logger.info(`Miner started. Assembling next block in ${this.period / 1000}s`)
+    return true
+  }
+
+  /**
+   * Assembles a block from txs in the TxPool and adds it to the chain.
+   * If a new block is received while assembling it will abort.
+   */
+  async assembleBlock() {
+    if (this.assembling) {
+      return
+    }
+    this.assembling = true
+
+    // Abort if a new block is received while assembling this block
+    let interrupt = false
+    const setInterrupt = () => {
+      interrupt = true
+      this.assembling = false
+    }
+    this.config.events.on(Event.CHAIN_UPDATED, setInterrupt.bind(this))
+
+    const parentBlockHeader = this.latestBlockHeader()
+    const number = parentBlockHeader.number.addn(1)
+    const { gasLimit } = parentBlockHeader
+    const [signerAddress, signerPrivKey] = this.config.accounts[0]
+
+    // Abort if we have too recently signed
+    if (this.config.chainCommon.consensusType() === ConsensusType.ProofOfAuthority) {
+      const header = BlockHeader.fromHeaderData(
+        { number },
+        { common: this.config.chainCommon, cliqueSigner: signerPrivKey }
```

Comment:
well, not really needed here because this is just constructing a header to use the function `blockchain.cliqueCheckRecentlySigned()` and I doubt any future hf will have PoA changes that affect the logic in deriving that answer.  It wouldn't hurt to add it though, so I will if you think it would be more future proof.

re: using `hardforkByTD` I'm hesitant to add merge logic to the miner yet since it's still a bit undecided, once everything is more formalized I can separately work on a PR for it. I'm sure there will be lots of questions for a miner in a post-merge world - but I guess its utility would be to provide the execution environment to return a valid block for an eth2 validator? Would PoA continue incognizant of the merge hardfork? Interesting times ahead!

Reply:
🙂